l10n_it_delivery_note/models/stock_delivery_note.py

```diff
@@ -709,110 +725,218 @@ def _check_delivery_notes_before_invoicing(self):
+            .with_context(lang=sale.partner_invoice_id.lang)
+            ._prepare_invoice()
+        )
+        return invoice_vals
+
+    def _prepare_invoice_lines(self, sale_orders, from_so, invoice_method, final):
+        """Creates invoice lines from delivery note lines,
+        sorting delivery notes by date and name.
+        For each delivery note adds:
+        - A section line with delivery note data
+        - Product lines from the delivery note
+        If called from a sale order (SO):
+        - Adds order lines not yet in delivery notes
+        If requested (invoice_method == "service"):
+        - Adds service type lines from orders
+        If it's the final invoice (final == True):
+        - Adds down payment lines
+        Sets the created lines in the invoice values dictionary"""
+        vals_list = []
+        sequence = 1
+        account_move = self.env["account.move"]
+
+        # Add delivery note lines as sections
+        for dn in self.sorted(key=lambda d: (d.date, d.name)):
+            vals_list.append(
+                Command.create(account_move._prepare_note_dn_value(sequence, dn))
+            )
+            sequence += 1
+
+            # Get delivery note lines
+            dn_line_ids = dn.line_ids
+            if from_so:
+                # Filter lines related to the SO if called from SO
+                dn_line_ids = dn_line_ids.filtered(
+                    lambda line, from_so=from_so: line.sale_line_id
+                    and line.sale_line_id.order_id.id == from_so
                 )
-            if delivery_note_id.state == "draft":
-                raise UserError(_("%s is in draft!") % delivery_note_id.display_name)
-            for line in delivery_note_id.line_ids:
-                if line.product_id.invoice_policy == "order":
-                    raise UserError(
-                        _(
-                            "In %(ddt_name)s there is %(product_name)s"
-                            " with invoicing policy 'order'"
+
+            # Add delivery note lines
+            for line in dn_line_ids:
+                vals = line._prepare_invoice_line(sequence=sequence)
+                vals_list.append(Command.create(vals))
+                sequence += 1
+
+        # Add remaining SO lines if called from SO
+        if from_so:
+            sale_lines = sale_orders.mapped("order_line").filtered(
+                lambda ol: not ol.delivery_note_line_ids
+                and ol.product_id.type != "service"
+                and ol.qty_to_invoice > 0
+            )
+            for line in sale_lines:
+                vals = line._prepare_invoice_line(sequence=sequence)
+                vals["sale_line_ids"] = [(4, line.id)]
+                vals_list.append(Command.create(vals))
+                sequence += 1
+
+        # Add service lines if requested
+        if invoice_method == "service":
+            sale_ids = sale_orders if from_so else sale_orders
```

Comment:
forse era  `sale_ids = sale_orders[0]`
altrimente è  `sale_ids = sale_orders` 
in ogni caso l'if non serve